Update for "Retriable and non-retriable Pod failures for Jobs" (#3646)
kubernetes/enhancements: keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

to do not add any pod condition in this case. It should be re-considered in the
future if there is a good motivating use-case.

##### Admission failures

In some scenarios a pod admission failure could be restarted and is likely to
succeed on another node (for example a pod scheduled to a node with resource
pressure, see: [Pod admission error](#pod-admission-error)). However, in some

> situations isn't clear as the failure can be caused by incompatible pod and node

Reviewer: How about something like: 

```
In some scenarios a pod admission failure could result in a successful pod restart on another 
node (for example a pod scheduled to a node with resource pressure, see: 
[Pod admission error](#pod-admission-error)). However, in other situations it won't be as clear, 
since the failure can be caused by incompatible pod and node configurations. Node configurations 
are often the same within a cluster, so it is likely that the pod would fail if restarted on any 
other node in the cluster. In that case, adding `DisruptionTarget` condition could cause 
a never-ending loop of retries, if the pod failure policy was configured to ignore such failures. 
Given the above, we decide not to add any pod condition for such failures. If there is a sufficient
motivating use-case, a dedicated pod condition might be introduced to annotate some of the 
admission failure scenarios.
```

configurations. As node configurations are often the same within a cluster it is
likely that the pod would fail if restarted on any node in the cluster. In that
case, adding `DisruptionTarget` condition could cause a never-ending loop of
retries if the pod failure policy was configured to ignore such failures. Thus,
we decide not to add any pod condition for such failures. If there is a
sufficient motivating use-case a dedicated pod condition might be introduced to
annotate some of the admission failure scenarios.

##### Resource limits exceeded

> A Pod failure-initiated by Kubelet can be caused by exceeding pod's

Reviewer: ```suggestion
A Pod failure initiated by Kubelet can be caused by exceeding pod's
```

(or container's) resource (memory or ephemeral-storage) limits. We have considered
(and prepared an initial implementation, see PR
[Add ResourceExhausted pod condition for oom killer and exceeding of local storage limits](https://github.com/kubernetes/kubernetes/pull/113436))
introduction of a dedicated Pod failure condition `ResourceExhausted` to
annotate pod failures due to the above scenarios.

However, it turned out, that there are complications with detection of exceeding

> of memory limits:

Reviewer: ```suggestion
memory limits:
```

- the approach we considered is to rely on the Out-Of-Memory (OOM) killer.
In particular, we could detect that a pod was terminated due to OOM killer based
on the container's `reason` field being equal to `OOMKilled`. This value is set
on Linux by the leading container runtime implementations: containerd (see
[here](https://github.com/containerd/containerd/blob/23f66ece59654ea431700576b6020baffe1a4e49/pkg/cri/server/events.go#L344)
for event handling and
[here](https://github.com/containerd/containerd/blob/36d0cfd0fddb3f2ca4301533a7e7dcf6853dc92c/pkg/cri/server/helpers.go#L62)
for the constant definition)
and CRI-O (see
[here](https://github.com/cri-o/cri-o/blob/edf889bd277ae9a8aa699c354f12baaef3d9b71d/server/container_status.go#L88-L89)).

> - setting of the `reason` field to equal `OOMKilled` is not standardized. We have

Reviewer: ```suggestion
- setting the `reason` field to `OOMKilled` is not standardized, either. We have
```

> started an effort to standardize the handling of handling OOM killed containers

Reviewer: ```suggestion
started an effort to standardize the handling of OOM killed containers
```

(see: [Documentation for the CRI API reason field to standardize the field for containers terminated by OOM killer](https://github.com/kubernetes/kubernetes/pull/112977)).

> However, in the process it turned out that on some configurations

Reviewer: ```suggestion
However, in the process it turned out that in some configurations
```

(for example the CRI-O with cgroupv2, see:
[Add e2e_node test for oom killed container reason](https://github.com/kubernetes/kubernetes/pull/113205)),
the container's `reason` field is not set to `OOMKilled`.
- OOM killer might get invoked not only when container's limits are exceeded,

> but also when the system is running low on memory. In such a scenario there

Reviewer: ```suggestion
but also when the system is running low on memory. In such scenario there
```

can be race conditions in which both the `DisruptionTarget` condition and the
`ResourceExhausted` could be added.

Thus, we decide not to annotate the scenarios with the `ResourceExhausted`
condition. While there are not known issues with detection of the exceeding of

> Pod's ephemeral storage limits we prefer to avoid future extension of the

Reviewer: ```suggestion
Pod's ephemeral storage limits, we prefer to avoid future extension of the
```

semantics of the new condition type. Alternatively, we could introduce a pair of
dedicated pod condition types: `OOMKilled` and `EphemeralStorageLimitExceeded`.
This approach, however, could create an unnecessary proliferation of the pod
condition types.

> Finally, we would like to first hear user feedback on the preferred approach

Reviewer: Maybe worth adding this as a possible extensions at the end of this KEP, since you are implying that option, just not atm.

Author: Added a small section in Alternatives

- DeletionByTaintManager (Pod evicted by kube-controller-manager due to taints)
- EvictionByEvictionAPI (Pod deleted by Eviction API)
- DeletionByPodGC (an orphaned Pod deleted by pod GC)
- TerminationByKubelet (Pod terminated due to graceful node shutdown or node resource pressure).

The already existing `status.conditions` field in Pod will be used by kubernetes
components to append a dedicated condition.

When the failure is initiated by a component which deletes the pod, then the API
call to append the condition will be issued as a pod status update call
before the Pod delete request (not necessarily as the last update request before
the actual delete). For Kubelet, which does not delete the pod itself, the pod

> condition is added in the same API request as to change the phase to failed.

Reviewer: ```suggestion
condition is added in the same API request as the phase change to failed.
```
